### PR TITLE
Upgrade Phi from Phi-3 mini to Phi-4 mini

### PR DESCRIPTION
**Pull request by Honry**
The Phi-4 Mini model is generated by ORT-GenAI.

This PR also fixes some other issues:

 - Fix long TTF (time to first token) issue for WebGPU backend
 - Use ort.webgpu.min.js for published ort-web, as which contains latest impelmentation of WebGPU EP. (WebNN EP is also included)
 - Use concat instead of push for padding array, to avoid RangeError for concating long arrays

### CODE REVIEW

File: demos/text-generation/llm.js [hunk 1]
```diff
@@ -245,13 +245,32 @@ export class LLM {
 
         let attnMask = Array.from({ length: inputIdsLen }, () => BigInt(1));
         const positionIds = Array.from({ length: inputIdsLen }, (_, i) => BigInt(i++));
-        // Padding input_ids, attention_mask, position_ids to have length of this.maxLength
-        const inputIdsBuffer = this.paddingInput(inputIds, this.maxLength);
+        // Both input_ids and position_ids have shapes of [batch_size, sequence_length].
+        // the sequence_length is the length of inputIds, which is dynamic.
```
Comment (fdwr) [inline demos/text-generation/llm.js:249]:
```suggestion
        // The sequence_length is the length of inputIds, which is dynamic.
```

File: demos/text-generation/llm.js [hunk 2]
```diff
@@ -245,13 +245,32 @@ export class LLM {
...
+        // Since WebNN does not support dynamic shapes, fix the sequence_length to maxLength and
```
Comment (fdwr) [inline demos/text-generation/llm.js:250]: > Since WebNN does not support dynamic shapes

Even if WebNN supports dynamic shapes in the future, the reality is that certain backends will need to regenerate the graph whenever an input size changes, and performance-wise, backends execute much faster when known sizes are passed for allocation planning. So it would probably still be worth it to pad them.

(resolve me - no action except to read)

File: demos/text-generation/llm.js [hunk 3]
```diff
@@ -245,13 +245,32 @@ export class LLM {
...
+        // pad the rest elements with 0 value.
+        // TODO: This may cause an overflow error if maxLength is excessively large,
+        // as it could exceed the allowable array size or memory limits.
+        // e.g. QWen2.0 supports max_length: 32768, in a matmul of the GQA decomposed op,
+        // its input shapes will be [1, 14, 32768, 64] x [1, 14, 64, 32768] = [1, 14, 32768, 32768]
+        // which exceeds the 2GB tensor size limitation.
```
Comment (fdwr) [inline demos/text-generation/llm.js:256]: Oof, entire models used to be < 100MB's, and now single tensors are passing the limits of 32-bit. It's no wonder that many ML libraries went with int64 for their indices. Related: https://github.com/microsoft/onnxruntime/issues/25445

> This happened in [chatGLM's](https://huggingface.co/THUDM/chatglm3-6b/blob/main/modeling_chatglm.py) mlp layer which has a weight martrix 4096 * 27392, then with batch size 256, the tensor elements will be 4096 * 27392 * 256 = 28,722,593,792 > 4,294,967,295.

(resolve me - no action except to read)

File: demos/text-generation/main.js [hunk 4]
```diff
@@ -34,20 +34,19 @@ const MODELS = {
...
+        id: "microsoft/Phi-4-mini-instruct-onnx",
+        remote_id: "microsoft/Phi-4-mini-instruct",
         file_name: "model.onnx",
-        local_path: "models/microsoft/directml-int4-awq-block-128/",
-        remote_path:
-            "https://huggingface.co/microsoft/Phi-3-mini-4k-instruct-onnx/resolve/main/directml/directml-int4-awq-block-128/",
-        eos_token_id: [32000, 32001, 32007],
-        max_length: 4096,
+        local_path: "models/microsoft/Phi-4-mini-instruct-onnx/",
+        remote_path: "https://huggingface.co/webnn/Phi-4-mini-instruct-onnx/resolve/main/onnx/",
```
Comment (fdwr) [inline demos/text-generation/main.js:44]: Just curious - how did you find this path? (the user facing path is `https://huggingface.co/webnn/Phi-4-mini-instruct-onnx/blob/main/onnx/model.onnx`, but replacing "blob" with "resolve" gets you a direct download)

**Review by fdwr — APPROVED** [review]
👍

File: assets/js/common_utils.js [hunk 5]
```diff
@@ -194,7 +194,7 @@ const loadScriptWithMessage = async version => {
             if (version === "latest") {
                 version = await getLatestOrtWebDevVersion();
             }
-            await loadScript("onnxruntime-web", `${ORT_CDN_URL}${version}/dist/ort.all.min.js`);
+            await loadScript("onnxruntime-web", `${ORT_CDN_URL}${version}/dist/ort.webgpu.min.js`);
```
[the remark this answers is not in the document]
Comment (Honry) [reply assets/js/common_utils.js:197]: It will finally be in "all".

File: demos/text-generation/llm.js [hunk 6]
```diff
@@ -163,12 +165,12 @@ export class LLM {
...
-            const empty =
-                this.dataType === "float16" ? new Float16Array(kvNumElements) : new Float32Array(kvNumElements);
+            // Initialize kv cache as empty tensors for webgpu or wasm EP
+            const emptyDims = [1, this.kvNumHeads, 0, this.headSize];
+            const empty = this.dataType === "float16" ? new Float16Array() : new Float32Array();
             for (let i = 0; i < this.numLayers; ++i) {
-                this.feed[`past_key_values.${i}.key`] = new ort.Tensor(this.dataType, empty, this.kvDims);
-                this.feed[`past_key_values.${i}.value`] = new ort.Tensor(this.dataType, empty, this.kvDims);
+                this.feed[`past_key_values.${i}.key`] = new ort.Tensor(this.dataType, empty, emptyDims);
+                this.feed[`past_key_values.${i}.value`] = new ort.Tensor(this.dataType, empty, emptyDims);
```
[the remark this answers is not in the document]
Comment (Honry) [reply demos/text-generation/llm.js]: Done.

**Review by fdwr — APPROVED** [review]
👍

**Comment by fdwr** [comment]
Sorry, I merged before @ibelem had a chance to look 😅.
